### ISSUE TITLE
libstdc++ should not be forced (vs some other C++ STL library) in library like OpenCV

### ISSUE THREAD

**Issue by seanm** [opencv/opencv#7566]
##### System information (version)
- OpenCV => 3.1 & current git master
- Operating System / Platform => Mac OS X
- Compiler => Xcode / clang
##### Detailed description

In cmake/OpenCVCompilerOptions.cmake there is the following block:

```
# Extra link libs if the user selects building static libs:
if(NOT BUILD_SHARED_LIBS AND CMAKE_COMPILER_IS_GNUCXX AND NOT ANDROID)
  # Android does not need these settings because they are already set by toolchain file
  set(OPENCV_LINKER_LIBS ${OPENCV_LINKER_LIBS} stdc++)
  set(OPENCV_EXTRA_FLAGS "-fPIC ${OPENCV_EXTRA_FLAGS}")
endif()
```

Historically, Mac OS X included GNU libstdc++, but they stopped updating it at version 4.2 (ancient) and it's deprecated in favour of llvm's libc++.

On other platforms too, there are choices of C++ STL libraries.

Forcing an STL library choice like this seems to me incorrect when building a _library_ like OpenCV. The problem is that the STL library choice must be the same for all code built, that is, my app, library1, library2, etc. must all be built with the same STL library, otherwise crazy link issues will ensue.

In my particular case, just removing "stdc++" worked, but I'm not sure if that's an actual solution.

**Comment by seanm** [comment]
Just a note to say this is still an issue in OpenCV 3.3.

**Comment by seanm** [comment]
Just a note to say this seems fixed in 3.4.  At least, "stdc++" is no longer in that file, and I can now build out-of-the-box.

**Comment by alalek** [comment]
#8198
